src/odbc/sql2tds.c

```diff
@@ -149,6 +149,171 @@ _odbc_blob_free(TDSCOLUMN *col)
+
+	if (fSqlType == SQL_DECIMAL || fSqlType == SQL_NUMERIC)
+		is_numeric = 1;
+
+	if (ipar > apd->header.sql_desc_count && desc_alloc_records(apd, ipar) != SQL_SUCCESS)
+		return TDS_FAIL;
+
+	drec = &apd->records[ipar - 1];
+
+	if (odbc_set_concise_c_type(fCType, drec, 0) != SQL_SUCCESS)
+		return TDS_FAIL;
+
+	if (drec->sql_desc_type == SQL_C_CHAR || drec->sql_desc_type == SQL_C_WCHAR || drec->sql_desc_type == SQL_C_BINARY)
+		drec->sql_desc_octet_length = cbValueMax;
+	drec->sql_desc_indicator_ptr = pcbValue;
+	drec->sql_desc_octet_length_ptr = pcbValue;
+	drec->sql_desc_data_ptr = (char *) rgbValue;
+
+	if (ipar > ipd->header.sql_desc_count && desc_alloc_records(ipd, ipar) != SQL_SUCCESS)
+		return TDS_FAIL;
+
+	drec = &ipd->records[ipar - 1];
+
+	drec->sql_desc_parameter_type = fParamType;
+	if (odbc_set_concise_sql_type(fSqlType, drec, 0) != SQL_SUCCESS)
+		return TDS_FAIL;
+
+	if (is_numeric) {
+		drec->sql_desc_precision = cbColDef;
+		drec->sql_desc_scale = ibScale;
+	} else {
+		drec->sql_desc_length = cbColDef;
+	}
+
+	return TDS_SUCCESS;
+}
+
+static TDS_INT
+odbc_convert_table(TDS_STMT * stmt, SQLTVP * src, TDS_TVP * dest, SQLLEN num_rows)
+{
+	SQLLEN i;
+	int j;
+	TDS_TVP_ROW * row;
+	TDS_TVP_ROW ** prow;
+	TDSPARAMINFO * params, * new_params;
+	TDS_DESC * apd, * ipd;
+	SQLPOINTER rgbValue;
+	SQLTVPCOLUMN ** cols;
+	SQLTVPCOLUMNLIST * node;
+	char * type_name, * pch;
+
+	dest->num_cols = src->num_cols;
+	dest->metadata = NULL;
+	dest->row = NULL;
+
+	if ((type_name = tds_strndup(src->type_name, src->type_name_len)) == NULL) {
+		odbc_errs_add(&stmt->errs, "HY001", NULL);
+		ODBC_EXIT_(stmt);
+	}
+
+	/* Tokenize and extract the schema & TVP typename from the TVP's full name */
+	pch = strchr(type_name, '.');
+	if (pch == NULL) {
+		dest->schema = tds_strndup("", 0);
+		dest->name = type_name;
+	} else {
+		*pch = 0;
+		dest->schema = tds_strndup(type_name, src->type_name_len);
+		dest->name = tds_strndup(++pch, src->type_name_len);
+		free(type_name);
+	}
+
+	/* Ensure that the TVP typename does not contain any more '.' */
+	/* Otherwise, report it as an invalid data type error */
+	pch = strchr(dest->name, '.');
+	if (pch != NULL) {
+		odbc_errs_add(&stmt->errs, "HY004", NULL);
+		ODBC_EXIT_(stmt);
+	}
+
+	/* Convert the columns from a linked list to an array */
+	if ((cols = malloc(src->num_cols * sizeof(SQLTVPCOLUMN *))) == NULL)
+		return -1;
+
+	for (node = src->col_list; node != NULL; node = node->next)
+		cols[node->ipar - 1] = node->column;
+
+	/* Create a dummy row to store column metadata */
+	apd = desc_alloc(stmt, DESC_APD, SQL_DESC_ALLOC_AUTO);
+	ipd = desc_alloc(stmt, DESC_IPD, SQL_DESC_ALLOC_AUTO);
+
+	if ((dest->metadata = malloc(sizeof(TDS_TVP_ROW))) == NULL)
+		return -1;
+
+	dest->metadata->next = NULL;
+	params = NULL;
+	for (j = 0; j < src->num_cols; j++) {
+		if (!(new_params = tds_alloc_param_result(params)))
+			return -1;
+
+		odbc_convert_table_row(apd, ipd, j + 1, cols[j]->fParamType, cols[j]->fCType,
```

Comment:
More than a row here we are converting a column. Also there's no error check although the function returns error.